File: Editorial/109_Convert_Sorted_List_to_Binary_Search_Tree/C/3_Inorder_Simulation.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 *
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct ListNode* head;
int findSize(struct ListNode* head) {
    struct ListNode* ptr = head;
    int c = 0;
    while (ptr != NULL) {
        ptr = ptr->next;
        c += 1;
    }
    return c;
}
struct TreeNode* convertListToBST(int l, int r) {
    if (l > r) return NULL;
    int mid = (l + r) / 2;
    struct TreeNode* left = convertListToBST(l, mid - 1);
    if (head == NULL) return NULL;
    struct TreeNode* node = malloc(sizeof(struct TreeNode));
    node->val = head->val;
    node->left = left;
    head = head->next;
    node->right = convertListToBST(mid + 1, r);
    return node;
}
struct TreeNode* sortedListToBST(struct ListNode* h) {
    head = h;
    int size = findSize(head);
    return convertListToBST(0, size - 1);
}

File: Editorial/0101-0200/0109_Convert_Sorted_List_to_Binary_Search_Tree/TypeScript/3_Inorder_Simulation.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 *
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function sortedListToBST(head: ListNode | null): TreeNode | null {
    let currentNode = head;
    let findSize = function (head: ListNode | null): number {
        let ptr = head,
            c = 0;
        while (ptr) {
            ptr = ptr.next;
            c++;
        }
        return c;
    };
    let convert = function (l: number, r: number): TreeNode | null {
        if (l > r) return null;
        let mid = Math.floor((l + r) / 2);
        let left = convert(l, mid - 1);
        let node = new TreeNode(currentNode.val);
        node.left = left;
        currentNode = currentNode.next;
        node.right = convert(mid + 1, r);
        return node;
    };
    let size = findSize(head);
    return convert(0, size - 1);
}
